R/civis_ml_plot.R

```diff
@@ -0,0 +1,115 @@
+
+#' Plot a histogram of the predicted scores or probabilities for a CivisML object
+#' @param x civis_ml object
+#' @param name name of the class (for multiclass output)
+#' @param ... unused.
+#' @return A \code{ggplot2} plot object is returned invisibly.
+#' @export
+hist.civis_ml <- function(x, name = NULL, ...) {
+  score_array <- get_metric(x, "score_histogram")
+  if (is_multiclass(x)) {
+    names <- get_model_data(x, "class_names")
+  } else {
+    names <- get_model_data(x, "target_columns")
+  }
+  if (is.null(score_array)) stop("Histogram data not available.")
+  if (is.array(score_array)) {
+    # 3d array to list of 2d arrays
+    score_array <- purrr::array_tree(score_array, margin = 1)
+  }
+  df <- data.frame(do.call(rbind, score_array))
+  colnames(df) <- c("low", "count", "up")
+  # if there is no rounding, bar widths -> 0 because of small differences between adjacent midpoints.
+  df$mp <- round((df$low + df$up)/2, 1)
+  df$freq <- df$count/sum(df$count)
+  df$names <- rep(names, sapply(score_array, nrow))
+  if (!is.null(name)) df <- subset(df, names == name)
+
+  ggplot2::ggplot(df) +
+    ggplot2::geom_bar(ggplot2::aes_string(x = "mp", y = "freq"), stat = "identity") +
+    ggplot2::xlab("OUT OF SAMPLE SCORES") +
+    ggplot2::ylab("DENSITY") +
+    ggplot2::scale_y_continuous(labels = percent) +
+    ggplot2::facet_grid(~names) +
+    ggplot2::theme_classic()
+}
+
+#' Y-yhat plot for regression with civis_ml
+#' @param x \code{civis_ml} object
+#' @param ... unused
+#' @return A \code{ggplot2} plot object is returned invisibly.
+#' @export
+plot.civis_ml_regressor <- function(x, ...) {
+  pl <- get_metric(x, "y_yhat_plot")
+  if (is.null(pl)) stop("Plotting data not available.")
+
+  # values is a matrix (binned histogram) with y values along the rows, and yhat values on the columns.
+  # this matrix is transformed from wide to long for plotting with ggplot2.
+  values <- data.frame(pl$values)
+  df <- utils::stack(values)
+  df$col_id <- as.numeric(df$ind)
+  df$row_id <- rep(1:nrow(values), ncol(values))
+
+  # valid y and yhat values are created from their ranges and step sizes.
+  y_vals <- round(seq(from = pl$y_range[1], to = pl$y_range[2], by = pl$y_step))
+  yhat_vals <- round(seq(from = pl$yhat_range[1], to = pl$yhat_range[2], by = pl$yhat_step))
+
+  df$x <- y_vals[df$row_id]
+  df$y <- yhat_vals[df$col_id]
+  ggplot2::ggplot(df, ggplot2::aes_string(x = "x", y = "y", fill = "values")) +
+    ggplot2::geom_tile() +
+    ggplot2::scale_fill_gradient("Bin Size", low = "white", high = civisblue) +
+    ggplot2::geom_abline(intercept = 0, slope = 1, color = civisyellow) +
+    ggplot2::ylab("Predictions") + ggplot2::xlab("Actuals") +
```

Comment:
Super minor point, but this chart has Proper Capitalization while the others are in ALL CAPS